internal-dns/types/src/v2/config.rs

```diff
@@ -0,0 +1,249 @@
+
+        v1::config::DnsConfigZone {
+            zone_name,
+            records: records
+                .into_iter()
+                .filter_map(|(k, v)| {
+                    let converted_records: Vec<v1::config::DnsRecord> =
+                        v.into_iter().filter_map(|rec| rec.as_v1()).collect();
+                    if converted_records.is_empty() {
+                        None
+                    } else {
+                        Some((k, converted_records))
+                    }
+                })
+                .collect(),
+        }
+    }
+}
+
+#[derive(
+    Clone,
+    Debug,
+    Serialize,
+    Deserialize,
+    JsonSchema,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+)]
+#[serde(tag = "type", content = "data")]
+pub enum DnsRecord {
+    A(Ipv4Addr),
+    // The renames are because openapi-lint complains about `Aaaa` and `Srv`
+    // not being in screaming snake case. `Aaaa` and `Srv` are the idiomatic
+    // Rust casings, though.
+    #[serde(rename = "AAAA")]
+    Aaaa(Ipv6Addr),
+    #[serde(rename = "SRV")]
+    Srv(Srv),
+    #[serde(rename = "NS")]
+    Ns(String),
+    #[serde(rename = "SOA")]
+    Soa(Soa),
+}
+
+impl DnsRecord {
+    fn as_v1(self) -> Option<v1::config::DnsRecord> {
+        match self {
+            DnsRecord::A(ip) => Some(v1::config::DnsRecord::A(ip)),
+            DnsRecord::Aaaa(ip) => Some(v1::config::DnsRecord::Aaaa(ip)),
+            DnsRecord::Srv(srv) => Some(v1::config::DnsRecord::Srv(srv.into())),
+            DnsRecord::Ns(_) | DnsRecord::Soa(_) => {
+                // V1 DNS records do not have variants for NS or SOA records, so
+                // we're lossy here.
+                None
+            }
+        }
+    }
+}
+
+// The `From<Ipv4Addr>` and `From<Ipv6Addr>` implementations are very slightly
+// dubious, because a v4 or v6 address could also theoretically map to a DNS
+// PTR record
+// (https://www.cloudflare.com/learning/dns/dns-records/dns-ptr-record/).
+// However, we don't support PTR records at the moment, so this is fine. Would
+// certainly be worth revisiting if we do in the future, though.
+
+impl From<Ipv4Addr> for DnsRecord {
+    fn from(ip: Ipv4Addr) -> Self {
+        DnsRecord::A(ip)
+    }
+}
+
+impl From<Ipv6Addr> for DnsRecord {
+    fn from(ip: Ipv6Addr) -> Self {
+        DnsRecord::Aaaa(ip)
+    }
+}
+
+impl From<Srv> for DnsRecord {
+    fn from(srv: Srv) -> Self {
+        DnsRecord::Srv(srv)
+    }
+}
+
+#[derive(
+    Clone,
+    Debug,
+    Serialize,
+    Deserialize,
+    JsonSchema,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+)]
+pub struct Srv {
+    pub prio: u16,
+    pub weight: u16,
+    pub port: u16,
+    pub target: String,
+}
+
+impl From<v1::config::Srv> for Srv {
+    fn from(other: v1::config::Srv) -> Self {
+        Srv {
+            prio: other.prio,
+            weight: other.weight,
+            port: other.port,
+            target: other.target,
+        }
+    }
+}
```

Comment:
the other option here is to use the `v1::config::Srv` type directly in v2, because it really has not changed. weaving the V1/V2 types together seems more difficult to think about generally, but i'm very open to the duplication being more confusing if folks feel that way.

Reply:
I would probably use the v1 types directly but I can see going either way.